### ISSUE TITLE
Simulation: Fix wrong scaling of nominals for inputs - [merged]

### ISSUE THREAD

**Issue by SGeeversAtVortech** [Deltares/rtc-tools#1380]
In GitLab by @vreeken on Feb 13, 2019, 23:11

_Merges fix-simulation-nominals-inputs -> master_

When setting up the rootfinder, we did not replace control/constant inputs
with their scaled equivalents in the DAE. We did however assume they were
scaled these when using set_var() to set the control inputs. This led to
horribly wrongly and inconsistently formulated problems.

Note that the initialize() method has its own optimization problem, and
there the constant and control inputs were replaced with their scaled
equivalents in the respective DAE.

To solve this issue for the rootfinder without breaking initialize(), we
now disallow/ignore nominals for constant inputs and control inputs. Note
that these variables are not part of the state vector, but are instead
parameters. Using unscaled variants in the DAE is therefore unlikely to
lead to issues.

* [x]  Merge !264 first

**Comment by SGeeversAtVortech** [comment]
In GitLab by @codecov on Feb 13, 2019, 23:26

# [Codecov](https://codecov.io/gl/deltares/rtc-tools/pull/260?src=pr&el=h1) Report
> Merging [#260](https://codecov.io/gl/deltares/rtc-tools/pull/260?src=pr&el=desc) into [master](https://codecov.io/gl/deltares/rtc-tools/commit/af32d81c7690cfe5549c09d073c8286cfe2ea453?src=pr&el=desc) will **increase** coverage by `0.01%`.
> The diff coverage is `100%`.


| [Impacted Files](https://codecov.io/gl/deltares/rtc-tools/pull/260?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [src/rtctools/simulation/simulation\_problem.py](https://codecov.io/gl/deltares/rtc-tools/pull/260/diff?src=pr&el=tree#diff-c3JjL3J0Y3Rvb2xzL3NpbXVsYXRpb24vc2ltdWxhdGlvbl9wcm9ibGVtLnB5) | `78.33% <100%> (+0.12%)` | :arrow_up: |

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Feb 15, 2019, 11:37

Commented on [src/rtctools/simulation/simulation_problem.py line 760](https://github.com/Deltares/rtc-tools/compare/091104e6f73f2dd1543facb2a081f3c6fb9e44c9..c39512dfa8eba882799c0a29da1a33793ba331f2#diff-5ea63c5cd21b27afb74ff666282a3367R760)



Actually, we could support input nominals just fine by not excluding them in line 180.
Doing so would be more robust than a comment forbidding it, imo.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Feb 15, 2019, 11:37

Commented on [src/rtctools/simulation/simulation_problem.py line 106](https://github.com/Deltares/rtc-tools/compare/af32d81c7690cfe5549c09d073c8286cfe2ea453..a5cedd8c73cbcb4fc1c0ef15a899a315214d6fdd#diff-5ea63c5cd21b27afb74ff666282a3367R106)



Like this, we also skip the variable types for inputs.  This needs to be addressed:  We will want to support boolean inputs, for example.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Feb 15, 2019, 13:00

Commented on [src/rtctools/simulation/simulation_problem.py line 760](https://github.com/Deltares/rtc-tools/compare/091104e6f73f2dd1543facb2a081f3c6fb9e44c9..c39512dfa8eba882799c0a29da1a33793ba331f2#diff-5ea63c5cd21b27afb74ff666282a3367R760)



I don't think it makes sense to support constant/control input nominals, as these are not decision variables (but parameters). 

It's not necessarily more robust, because some parts of the SimulationProblem code rely on keys _and_ values in `__nominals`, other parts only on keys of `__nominals` (and get values by calling `get_variable_nominal`), and even other just always call `get_variable_nominal`. Simply put, SimulationProblem is somewhat API stable, but fairly broken internally.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Feb 15, 2019, 13:02

Commented on [src/rtctools/simulation/simulation_problem.py line 760](https://github.com/Deltares/rtc-tools/compare/091104e6f73f2dd1543facb2a081f3c6fb9e44c9..c39512dfa8eba882799c0a29da1a33793ba331f2#diff-5ea63c5cd21b27afb74ff666282a3367R760)



Sorry, with robust I meant "less prone to user error".

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Feb 15, 2019, 13:02

Commented on [src/rtctools/simulation/simulation_problem.py line 106](https://github.com/Deltares/rtc-tools/compare/af32d81c7690cfe5549c09d073c8286cfe2ea453..a5cedd8c73cbcb4fc1c0ef15a899a315214d6fdd#diff-5ea63c5cd21b27afb74ff666282a3367R106)



Good catch. I guess it's better to move this `self.__python_types[sym_name] = v.python_type` outside this loop, such that it also works on parameters, der_states, and time.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Feb 15, 2019, 13:04

Commented on [src/rtctools/simulation/simulation_problem.py line 760](https://github.com/Deltares/rtc-tools/compare/091104e6f73f2dd1543facb2a081f3c6fb9e44c9..c39512dfa8eba882799c0a29da1a33793ba331f2#diff-5ea63c5cd21b27afb74ff666282a3367R760)



OK, fair point. I'll check if maybe changing the comment is enough. It might just be the case that the nominal on inputs is now ignored outright.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Feb 15, 2019, 13:05

marked as a **Work In Progress**

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Feb 16, 2019, 14:15

Commented on [src/rtctools/simulation/simulation_problem.py line 760](https://github.com/Deltares/rtc-tools/compare/091104e6f73f2dd1543facb2a081f3c6fb9e44c9..c39512dfa8eba882799c0a29da1a33793ba331f2#diff-5ea63c5cd21b27afb74ff666282a3367R760)



changed this line in [version 2 of the diff](https://gitlab.com/deltares/rtc-tools/merge_requests/260/diffs?diff_id=34218320&start_sha=c39512dfa8eba882799c0a29da1a33793ba331f2#15e34bb5e5838c1e165ddfcd99c9c9596b965eb6_760_760)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Feb 16, 2019, 14:15

added 1 commit

<ul><li>5a026edf - Simulation: Fix wrong scaling of nominals for inputs</li></ul>

[Compare with previous version](https://gitlab.com/deltares/rtc-tools/merge_requests/260/diffs?diff_id=34218320&start_sha=c39512dfa8eba882799c0a29da1a33793ba331f2)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Feb 16, 2019, 14:16

Commented on [src/rtctools/simulation/simulation_problem.py line 760](https://github.com/Deltares/rtc-tools/compare/091104e6f73f2dd1543facb2a081f3c6fb9e44c9..c39512dfa8eba882799c0a29da1a33793ba331f2#diff-5ea63c5cd21b27afb74ff666282a3367R760)



As per your suggestion, I fixed `set_var` and `get_var` to not apply nominals for non-decision variables. The comment is now removed. If someone would implement their own `get_variable_nominal` and return a value for a control/constant input, it would just be ignored (read: never looked up).

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Feb 16, 2019, 14:27

added 9 commits

<ul><li>5a026edf...6116b456 - 7 commits from branch <code>master</code></li><li>8bd0ffe7 - Simulation: Store variable Python type for all variables</li><li>4e689e74 - Simulation: Fix wrong scaling of nominals for inputs</li></ul>

[Compare with previous version](https://gitlab.com/deltares/rtc-tools/merge_requests/260/diffs?diff_id=34218527&start_sha=5a026edf0ab963bb8a687e592a608d5c6bef028b)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Feb 16, 2019, 14:27

unmarked as a **Work In Progress**

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Feb 16, 2019, 14:28

Commented on [src/rtctools/simulation/simulation_problem.py line 106](https://github.com/Deltares/rtc-tools/compare/af32d81c7690cfe5549c09d073c8286cfe2ea453..a5cedd8c73cbcb4fc1c0ef15a899a315214d6fdd#diff-5ea63c5cd21b27afb74ff666282a3367R106)



Addressed in !264 (should be merged first)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Feb 16, 2019, 14:28

resolved all discussions

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Feb 16, 2019, 15:59

I like this solution much better.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Feb 17, 2019, 11:49

added 2 commits

<ul><li>89116675 - Simulation: Store variable Python type for all variables</li><li>20ff4e10 - Simulation: Fix wrong scaling of nominals for inputs</li></ul>

[Compare with previous version](https://gitlab.com/deltares/rtc-tools/merge_requests/260/diffs?diff_id=34247567&start_sha=4e689e74b0029ac7a4891a8f91626915ea22f593)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Feb 17, 2019, 12:32

added 1 commit

<ul><li>efc16561 - Simulation: Fix wrong scaling of nominals for inputs</li></ul>

[Compare with previous version](https://gitlab.com/deltares/rtc-tools/merge_requests/260/diffs?diff_id=34248207&start_sha=20ff4e10e901364b76694067417cdcf0569a045f)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @baayen on Feb 18, 2019, 09:34

approved this merge request

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Feb 18, 2019, 10:47

added 2 commits

<ul><li>af32d81c - Simulation: Store variable Python type for all variables</li><li>a5cedd8c - Simulation: Fix wrong scaling of nominals for inputs</li></ul>

[Compare with previous version](https://gitlab.com/deltares/rtc-tools/merge_requests/260/diffs?diff_id=34291259&start_sha=efc16561360f0a16574230312b360d366f94718f)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Feb 18, 2019, 10:59

marked the checklist item **Merge !264 first** as completed

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Feb 18, 2019, 11:00

enabled an automatic merge when the pipeline for a5cedd8c73cbcb4fc1c0ef15a899a315214d6fdd succeeds

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Feb 18, 2019, 11:03

merged
